blockchain/blockchain.go

```diff
@@ -519,43 +519,29 @@ func (b *Blockchain) WriteHeadersWithBodies(headers []*types.Header) error {
 	return nil
 }
 
-// WriteBlocks writes a batch of blocks
-func (b *Blockchain) WriteBlocks(blocks []*types.Block) error {
-	// Check the size
-	size := len(blocks)
-	if size == 0 {
-		return fmt.Errorf("the passed in block array is empty")
+// WriteBlock writes a single block
+func (b *Blockchain) WriteBlock(block *types.Block) error {
+	// Check the param
+	if block == nil {
+		return fmt.Errorf("the passed in block is empty")
 	}
 
 	// Log the information
-	if size == 1 {
-		b.logger.Info(
-			"write block",
-			"num",
-			blocks[0].Number(),
-			"parent",
-			blocks[0].ParentHash(),
-		)
-	} else {
-		b.logger.Info(
-			"write blocks",
-			"num",
-			size,
-			"from",
-			blocks[0].Number(),
-			"to",
-			blocks[size-1].Number(),
-			"parent",
-			blocks[0].ParentHash())
-	}
-
-	parent, ok := b.readHeader(blocks[0].ParentHash())
+	b.logger.Info(
+		"write block",
+		"num",
+		block.Number(),
+		"parent",
+		block.ParentHash(),
+	)
+
+	parent, ok := b.readHeader(block.ParentHash())
```

Comment:
this is incurring a performance cost when syncing blocks. in the previous implementation we only checked if the first block's parent is present in the db. now we will check it all the time. 

can you benchmark the syncing tests with and without this line here just to see how much of a performance cost can we expect? 

do you see any simple and clean way we can skip this check when bulk syncing subsequent  blocks?

Reply:
You're completely right to assume it adds overhead - but that's something that we can only see when bulk syncing. Every other part of the PSDK already did this call on a per-block basis as they write 1 block at a time.

I don't see an easy way of skipping this check with the `WriteBlock` functionality, as it is a critical chain sanity check, it needs to be there at one point before pushing a block to the chain.

The header will be initially fetched from the in-memory cache and if there's a cache miss, it will grab it from LevelDB, either way it's a quick lookup as it's fetching only the header and not the entire block